phac-nml/irida
# Hotfix/collection date

## Description of changes

### Description 

This fixes an issue where, upon synchronizing a remote project, the **collectionDate** could be up to a day off from the original.  This makes two changes to fix the issue:

1. Adds `@JsonSerialize` and `@JsonDeserialize` annotations around **Sample.collectionDate**.
2. Adds a custom Json serialize/deserialize class `ca.corefacility.bioinformatics.irida.web.controller.api.json.DateJson` which should serialize/deserialize to the correct format **YYYY-MM-DD** (which was the format that was used previously).

Note: This is a `hotfix` and so should be merged into `master` and `development`.

### Testing

1. Create a new Project containing a new Sample
2. Set the **Collection Date** for the sample.
3. Setup a new Client and Remote Connection
4. Synchronize the Project you created.
5. Check the **Collection Date** for the new synchronized project, either in the UI or in the database.

The expected results will be:

#### Development/Master

1. Local Sample

    ![image](https://user-images.githubusercontent.com/200517/51703278-933c1d00-1fdb-11e9-8970-6f5a5f949b81.png)

2. Remote Sample

    ![image](https://user-images.githubusercontent.com/200517/51703372-cc748d00-1fdb-11e9-98e8-1d4db9d5c2dc.png)

3. `Sample` (local, remote) in Database

    ```sql
    select id,sampleName,remote_status,collectionDate from sample where id in (29,30);
    +----+------------+---------------+----------------+
    | id | sampleName | remote_status | collectionDate |
    +----+------------+---------------+----------------+
    | 29 | SampleDev  |          NULL | 2019-01-24     |
    | 30 | SampleDev  |            49 | 2019-01-23     |
    +----+------------+---------------+----------------+
    ```

#### This branch (with fix)

1. Local Sample

    ![image](https://user-images.githubusercontent.com/200517/51703211-67209c00-1fdb-11e9-8b53-ffe583717480.png)

2. Remote Sample

    ![image](https://user-images.githubusercontent.com/200517/51703404-df875d00-1fdb-11e9-9898-b1e403d8e873.png)

3. `Sample` (local, remote) in Database

    ```sql
    select id,sampleName,remote_status,collectionDate from sample where id in (29,30);
    +----+------------+---------------+----------------+
    | id | sampleName | remote_status | collectionDate |
    +----+------------+---------------+----------------+
    | 29 | SampleDev  |          NULL | 2019-01-24     |
    | 30 | SampleDev  |            49 | 2019-01-24     |
    +----+------------+---------------+----------------+
    ```

## Related issue

Fixes #256 

Issue #283 is related, but not fixed by this merge.

## Checklist

Things for the developer to confirm they've done before the PR should be accepted:
